Send email notification with smtplib.

adrianmoloce opened the pull request:
- Send email notification if send_mail flag is True
- Created config file for environment variables
- Created unit tests for scenarios where send email is successful and when sending it encounters errors.
- Add python-dotenv and pytest-mock as dependencies.

amoloce committed:
feat: Add pytest-mock and python-dotenv as dependencies

amoloce committed:
feat: Created config file with environment variables

amoloce committed:
feat: Send email notification with smtplib if send_mail

- Send email with smtplib if send_mail flag is true
- Created test for sending the mail and smtplib exceptions
- Created custom InvoiceRequestEmailError

koosie0507 commented on src/invoice_utils/config.py:
```diff
@@ -0,0 +1,11 @@
+
+load_dotenv()
+
+MAIL_HOST = os.getenv("MAIL_HOST")
```
It's always nice to namespace the names of the env vars with your app name. The global var's name can remain the same, but the env var should be namespaced. You can use pretty much anything as the namespace as long as there isn't a high chance of name collision. Longer strings have lower chances of colliding. E.g. `MAIL_HOST` -> `INVOICE_UTILS_MAIL_HOST`

koosie0507 commented on src/invoice_utils/config.py:
```diff
@@ -0,0 +1,11 @@
+load_dotenv()
+
+MAIL_HOST = os.getenv("MAIL_HOST")
+MAIL_PORT = os.getenv("MAIL_PORT")
```
`os.getenv` allows returning default values if the env var isn't present. Might want to use this feature for some of these variables.

koosie0507 commented on tests/test_web.py:
```diff
@@ -26,21 +31,51 @@ def test_send_mail_param_not_provided(http, caplog, invoice_request_body):
     invoice_request_body["address"] = "test@email.com"
+
+    mock_smtp = mocker.MagicMock(name="invoice_utils.web.smtplib.SMTP")
+    mocker.patch("invoice_utils.web.smtplib.SMTP", new=mock_smtp)
```
Does `invoice_utils.web.smtplib.SMTP` get un-patched at the end of running the test?

adrianmoloce replied:
It's a feature of pytest-mock, it automatically undoing the mocking at the end of the test.

koosie0507 commented on tests/test_web.py:
```diff
@@ -26,21 +31,51 @@ def test_send_mail_param_not_provided(http, caplog, invoice_request_body):
+    http, caplog, invoice_request_body, mocker
+):
     invoice_request_body["send_mail"] = True
     invoice_request_body["address"] = "test@email.com"
```
Remind me to talk to you about the [4 rules of simple design](https://martinfowler.com/bliki/BeckDesignRules.html) next time we meet (online or in person).

koosie0507 commented on tests/test_web.py:
```diff
@@ -26,21 +31,51 @@ def test_send_mail_param_not_provided(http, caplog, invoice_request_body):
     assert res.status_code == 201
+
+
+def test_send_mail_fails_for_smtp_exception(http, caplog, invoice_request_body, mocker):
```
Coupled with the commit structure, this doesn't look like the outcome of TDD. Should've had one assertion per test and should've started to extract common setup behaviour for the multiple tests that use it.
Next time, try limiting yourself to only one assertion per test (that fails)

amoloce committed:
feat: Add note for setting up env variables in README

amoloce committed:
feat: Add namespace and default values to environment variables.

koosie0507 commented on src/invoice_utils/config.py:
```diff
@@ -0,0 +1,14 @@
+INVOICE_UTILS_MAIL_LOGIN_USER = os.getenv("INVOICE_UTILS_MAIL_LOGIN_USER")
+INVOICE_UTILS_MAIL_LOGIN_PASSWORD = os.getenv("INVOICE_UTILS_MAIL_LOGIN_PASSWORD")
+INVOICE_UTILS_SENDER_EMAIL = os.getenv("INVOICE_UTILS_SENDER_EMAIL")
+INVOICE_UTILS_MAIL_SUBJECT = "Invoice generated with invoice-utils"
```
`os.environ["INVOICE_UTILS_MAIL_SUBJECT"] = ""`

amoloce committed:
Feat: add basic configurations for sending emails. (#7)

* feat: Config mail subject

* feat: SMTP login

- log into SMTP

* feat: Make SMTP Login Optional

* feat: Introduce SMTP start_tls optional env flag

* fix: Increase resilience of smtp tls flag

- method to convert string to bool for any env flag
- added tests cases for the tls flag
